tests/phpunit/tests/bookmark/getBookmark.php

```diff
@@ -0,0 +1,173 @@
+		if ( ARRAY_A === $output || ARRAY_N === $output ) {
+			$this->assertInternalType( 'array', $actual );
+			$this->assertSame( $expected, $actual );
+		} else {
+			$this->assertInstanceOf( stdClass::class, $actual );
+			$this->assertEquals( $expected, $actual );
+		}
+		$this->assertFalse( $this->query_called );
+	}
+
+	public function test_should_set_link_category_to_term_tax_id() {
+		$expected      = $this->bookmark;
+		$link_category = self::factory()->term->create(
+			array(
+				'taxonomy' => 'link_category',
+			)
+		);
+		$term_tax_id   = wp_add_object_terms( $this->bookmark->link_id, $link_category, 'link_category' );
+		if ( is_array( $term_tax_id ) ) {
+			$term_tax_id[0] = (string) $term_tax_id[0];
+		}
+		$expected->link_category = $term_tax_id;
+
+		$actual = get_bookmark( $this->bookmark->link_id );
+		$this->assertEquals( $expected, $actual );
+	}
+
+	public function test_should_set_link_category_to_null_and_throw_warning_when_tax_not_exist() {
+		// Store original taxonomy registration.
+		$taxonomy = $GLOBALS['wp_taxonomies']['link_category'];
+
+		// Remove the taxonomy to trigger the error.
+		unset( $GLOBALS['wp_taxonomies']['link_category'] );
+		$this->assertFalse( taxonomy_exists( 'link_category' ) );
+
+		$this->expectException( 'PHPUnit_Framework_Error_Warning' );
+		$this->expectExceptionMessage( 'Invalid taxonomy.' );
+
+		$actual = get_bookmark( $this->bookmark->link_id );
+		$this->assertNull( $actual->link_category );
+
+		// Restore the taxonomy.
+		$GLOBALS['wp_taxonomies']['link_category'] = $taxonomy;
+	}
+
+	/**
+	 * @param int|stdClass      $bookmark ID or instance of the bookmark.
+	 * @param array|object|null $expected Type returned depends on $output value.
+	 * @param string            $output   The required return type.
+	 * @param string            $filter   How to sanitize bookmark fields.
+	 */
+	private function run_bookmark_test( $bookmark, $expected, $output, $filter ) {
+		$actual = get_bookmark( $bookmark, $output, $filter );
+		$this->assertEquals( $expected, $actual );
+
+		if ( ARRAY_A === $output || ARRAY_N === $output ) {
+			$this->assertTrue( is_array( $actual ) );
+			$this->assertSame( $expected, $actual );
+			$key = ARRAY_N === $output ? $actual[0] : $actual['link_id'];
+		} else {
+			$this->assertInstanceOf( stdClass::class, $actual );
+			$this->assertEquals( $expected, $actual );
+			$key = $actual->link_id;
+		}
+
+		if ( ! isset( $GLOBALS['link'] ) ) {
+			$cached = wp_cache_get( $key, 'bookmark' );
+			$this->assertInstanceOf( stdClass::class, $cached );
+			$this->assertEquals( $this->bookmark, $cached );
+		}
+	}
+
+	public function data_bookmark() {
+		return array(
+			array( '', 'raw' ),
+			array( 'invalid', 'raw' ),
+			array( OBJECT, 'raw' ),
+			array( OBJECT, 'display' ),
+			array( ARRAY_A, 'raw' ),
+			array( ARRAY_A, 'display' ),
+			array( ARRAY_N, 'raw' ),
+			array( ARRAY_N, 'display' ),
+		);
+	}
+
+	public function get_expected( $output ) {
```

Comment:
```suggestion
	protected function get_expected( $output ) {
```

The helper function is only needed for and called from within this class, so can be made `protected` or even `private`. No need for it to be `public`.